elastic/kibana
# Improve state sync error handling

## Summary

Fixes https://github.com/elastic/kibana/issues/71461 regression since 7.7

New state syncing utils didn't properly handle errors,
Errors happening during URL parsing or writing wasn't handled, so state syncing could stop or in worth case blow out. (see #71461) 

There are not much scenarios where missing proper error handling could really impact users, except the one described in #71461: 

1. Kibana users `state:storeInSessionStorage`
2. Users often intuitively share hashed dashboard urls directly 
3. When someone opens those urls - there is a blank screen with warning

In 7.6 - dashboard would still load with default state. 
Since 7.7 these still could be achieved by removing query params for URL, but it is not obvious for regular users. 

This PR makes sure that behaviour is similar to one we had before 7.7.

### How to test 

Simplest way to test is to repeat real user flow:

1. Enabled `state:storeInSessionStorage`
2. Go to dashboard 
3. Copy url and open it in new incognito tab or different browser 
4. Dashboard should load. Warning(s) should be displayed. State (like time range or filters) should reset. 

### Please note

When testing you may notice multiple warnings, 
This is a separate common issue which needs attention: 
tracked here: https://github.com/elastic/kibana/issues/67270


### Implementation details 

After reviewing all state syncing usages, I decided to add error handling level not on a `state_sync` util level itself, but on storage level instead. You could view documentation diffs for insights. 

The reason is that `kbnUrlStateStorage` is used  both by applications code directly and indirectly by `state_sync` utils. If I'd add error handling apis on `state_sync` level, then error handling in application code would be more tedious: both had to make sure `state_sync` is setup to handle errors, but also every usage of `kbnUrlStateStorage` api is wrapped in try / catch.

### Release notes

Fix apps break on unrestorable session state in URL


### Checklist
